### PR TITLE
GLTFLoader Cubic Spline interpolation full support

### PR DESCRIPTION
**Pull request by takahirox**
See #12885 

Especially

https://github.com/mrdoob/three.js/pull/12885#issuecomment-351991718
https://github.com/mrdoob/three.js/pull/12885#issuecomment-352028202
https://github.com/mrdoob/three.js/pull/12885#issuecomment-352171507

Specification
https://github.com/KhronosGroup/glTF/blob/master/specification/2.0/README.md#appendix-c-spline-interpolation

/cc @donmccurdy

### CODE REVIEW

File: examples/js/loaders/GLTFLoader.js [hunk 1]
```diff
@@ -2361,12 +2392,31 @@ THREE.GLTFLoader = ( function () {
...
+							// Here is the trick to enable custom interpolation.
+							// Overrides .createInterpolant in a factory method which creates custom interpolation.
+							if ( sampler.interpolation === 'CUBICSPLINE' ) {
+
+								track.createInterpolant = function ( result ) {
+
+									// The keyframes of a cubic spline in glTF have input and output values where
+									// each input value corresponds to three output values. Then 3rd arguments which
+									// specifies sampleSize needs to be this.getValueSize() / 3 where this.getValueSize()
+									// is this.values.length / this.times.length.
```
Comment (donmccurdy) [inline examples/js/loaders/GLTFLoader.js]: I would rephrase this a little:

> A CUBICSPLINE keyframe in glTF has three output values for each input value, representing `inTangent`, `splineVertex`, and `outTangent`. As a result, `track.getValueSize()` must be divided by three to get the interpolant's `sampleSize` argument.

File: examples/js/loaders/GLTFLoader.js [hunk 2]
```diff
@@ -858,6 +858,64 @@ THREE.GLTFLoader = ( function () {
...
+
+	// Spline Interpolation
+	// Specification: https://github.com/KhronosGroup/glTF/blob/master/specification/2.0/README.md#appendix-c-spline-interpolation
+	function GLTFCubicSplineInterpolant( parameterPositions, sampleValues, sampleSize, resultBuffer ) {
+
+		THREE.Interpolant.call( this, parameterPositions, sampleValues, sampleSize, resultBuffer );
+
+	};
+
+	GLTFCubicSplineInterpolant.prototype = Object.assign( Object.create( THREE.Interpolant.prototype ), {
```
Comment (donmccurdy) [inline examples/js/loaders/GLTFLoader.js]: `Object.assign` is another thing that IE11 users would need to polyfill, since this file isn't compiled. May be better just to do

```js
GLTFCubicSplineInterpolant.prototype = Object.create( THREE.Interpolant.prototype );
GLTFCubicSplineInterpolant.prototype.constructor = GLTFCubicSplineInterpolant;
GLTFCubicSplineInterpolant.prototype.interpolate_ = function () { ... };
```

**Comment by takahirox** [comment]
Updated. Thanks for the comments.

> if you're able to make one in Blender that'd be great

Well, tbh I'm still very new to Blender...

> The ~three samples in the previous thread are working OK?

They look ok. The animations of two of them are very close to the video posted in that thread. The other one is very close to Animation in Blender (cube.blend).

**Comment by donmccurdy** [comment]
Sorry for the delay on this, I just don't have a model that shows use of the tangents in an obvious way, and I don't have access to Blender right now to create one... will review when i'm back in SF (~1.5 weeks from now).

**Comment by donmccurdy** [comment]
I'm back, and was able to put together an example with custom tangents:

[InterpolationComparison.zip](https://github.com/mrdoob/three.js/files/1602560/InterpolationComparison.zip)

The four boxes have the same three keyframe positions, but different interpolation. From left to right: `STEP`, `LINEAR`, `CUBICSPLINE`, `CUBICSPLINE` w/ tangents.

Results:

| Expected | Actual |
|---|---|
| interpolationcomparison | 2772c720-6c3d-438a-ace1-6f5e0f43d626-68822-0000960c57a2d904 |

Note that the yellow cube (far right) does not follow its tangents as expected. The "expected" GIF is created in Blender, and I'm not aware of another renderer that has implemented CUBICSPLINE+tangents yet, so it's entirely possible that the Blender export has issues here as well.

**Comment by takahirox** [comment]
Have you used the latest glTF-Blender-Exporter?

I export glTF with the latest exporter here and it became closer to the Blender animation.

ezgif com-optimize

Yeah, still a bit far from the Blender one but I think this's exporter's bug.
I'm gonna report to glTF-Blender-Exporter later.

**Review by donmccurdy — APPROVED** [review]
Awesome work! I did have an older version of the exporter, yes. This looks good to me, if the remaining mismatch does seem to be on Blender's side.

**Comment by mrdoob** [comment]
Thanks!
